### PR TITLE
docs(middleware-factory): snippets split, improved, and lint

### LINKED ISSUE #1367: Refactor Middleware factory docs code snippets

### PR DESCRIPTION
**Pull request by leandrodamascena**
**Issue number:** #1367 

### Changes

Changes snippet fictitious names to real code snippet filenames for easier discovery and collaboration to improve/correct document snippets.

Some examples have also been refactored so that the codes are complete and there are no lint/pre-commit hooks issues.
### User experience

Before:
image
image

After:
image
image

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

-----
[View rendered docs/utilities/middleware_factory.md](https://github.com/leandrodamascena/aws-lambda-powertools-python/blob/chore/middleware-factory/docs/utilities/middleware_factory.md)

### CODE REVIEW

File: docs/utilities/middleware_factory.md [hunk 1]
```diff
@@ -3,13 +3,22 @@ title: Middleware factory
 description: Utility
 ---
 
+<!-- markdownlint-disable MD043 -->
```
Comment (heitorlessa) [inline docs/utilities/middleware_factory.md:6]: This is actually our time to not disable this rule and implement `Getting started` and `Advanced` ;)

File: docs/utilities/middleware_factory.md [hunk 2]
```diff
@@ -3,13 +3,22 @@ title: Middleware factory
...
 * Trace each middleware when requested
 
+## Getting started
+
+???+ tip
+    All examples shared in this documentation are available within the [project repository](https://github.com/awslabs/aws-lambda-powertools-python/tree/develop/examples){target="_blank"}.
+
+You might need a middleware factory to abstract non-functional code and focus on business logic or even validate the payload before the lambda run, among other cases.
+
 ## Middleware with no params
```
Comment (heitorlessa) [inline docs/utilities/middleware_factory.md]:
```suggestion
### Middleware with no params
```

File: docs/utilities/middleware_factory.md [hunk 3]
```diff
@@ -18,74 +27,87 @@ You can create your own middleware using `lambda_handler_decorator`. The decorat
...
+    --8<-- "examples/middleware_factory/src/getting_started_middleware_before_logic_function.py"
+    ```
+
+=== "getting_started_middleware_before_logic_payload.json"
+
+    ```json hl_lines="9-13"
+    --8<-- "examples/middleware_factory/src/getting_started_middleware_before_logic_payload.json"
+    ```
+
+### Creating your own middleware for after logic
```
Comment (heitorlessa) [inline docs/utilities/middleware_factory.md]:
```suggestion
### Middleware for after logic
```

We can shorten this to ease navigation

File: docs/utilities/middleware_factory.md [hunk 4]
```diff
@@ -18,74 +27,87 @@ You can create your own middleware using `lambda_handler_decorator`. The decorat
 * **event** - Lambda function invocation event
 * **context** - Lambda function context object
 
-```python hl_lines="3-4 10" title="Creating your own middleware for before/after logic"
-from aws_lambda_powertools.middleware_factory import lambda_handler_decorator
+### Creating your own middleware for before logic
```
Comment (heitorlessa) [inline docs/utilities/middleware_factory.md]:
```suggestion
### Middleware with before logic
```

We can shorten this to ease navigation

File: docs/utilities/middleware_factory.md [hunk 5]
```diff
@@ -18,74 +27,87 @@ You can create your own middleware using `lambda_handler_decorator`. The decorat
...
 
-@obfuscate_sensitive_data(fields=["email"])
-def lambda_handler(event, context):
-	...
-```
+    ```json hl_lines="18 19 20"
+    --8<-- "examples/middleware_factory/src/getting_started_middleware_with_params_payload.json"
+    ```
 
 ## Tracing middleware execution
```
Comment (heitorlessa) [inline docs/utilities/middleware_factory.md]: This should be under `Advanced`

File: docs/utilities/middleware_factory.md [hunk 6]
```diff
@@ -18,74 +27,87 @@ You can create your own middleware using `lambda_handler_decorator`. The decorat
...
+    --8<-- "examples/middleware_factory/src/advanced_middleware_tracer_function.py"
+    ```
+
+=== "advanced_middleware_tracer_payload.json"
+
+    ```json
+    --8<-- "examples/middleware_factory/src/advanced_middleware_tracer_payload.json"
+    ```
+
+![Middleware avanced Tracer](../media/middleware_factory_tracer_2.png)
```
Comment (heitorlessa) [inline docs/utilities/middleware_factory.md:144]: Could you create another one `Combining Powertools utilities` section? 

One common use case is for customers to create a single middleware with several Powertools features like Tracer, Logger, Validation, etc.

**Review by heitorlessa — CHANGES_REQUESTED** [review]
Looks good too! Navigation needs fixing and made a suggestion to raise the bar by including a new section `Combining Powertools utilities`.

This should address a common question on how to best handle decorator stacking or create their own middleware with common Powertools utilities baked in

**Comment by leandrodamascena** [comment]
Hi @heitorlessa and @rubenfonseca.. Could you please review the example?

I think it covers a lot of features/utilities provided by Lambda Powertools and it is a real case for many customers. 

Thank you :smile:

File: examples/middleware_factory/sam/combining_powertools_utilities_template.yaml [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+      Policies:
+      - AWSLambdaBasicExecutionRole # Managed Policy
+      - Version: '2012-10-17' # Policy Document
+        Statement:
+          - Effect: Allow
+            Action:
+              - dynamodb:PutItem
+            Resource: !GetAtt HistoryTable.Arn
+          - Effect: Allow
+            Action: # https://docs.aws.amazon.com/appconfig/latest/userguide/getting-started-with-appconfig-permissions.html
```
Comment (leandrodamascena) [inline examples/middleware_factory/sam/combining_powertools_utilities_template.yaml:35]: I'm not sure if I'm comfortable with this, but from what I can see it's a real example in an official AWS documentation.

File: examples/middleware_factory/sam/combining_powertools_utilities_template.yaml [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        {
+              "save_history": {
+                "default": true
+              }
+        }
+      ContentType: 'application/json'
+
+  # this is just an example
+  # change this values according your deployment strategy
+  BasicDeploymentStrategy:
```
Comment (leandrodamascena) [inline examples/middleware_factory/sam/combining_powertools_utilities_template.yaml:118]: I added the deployment strategy to avoid the default baking time.. The default time (10 minutes) is too long to someone that wants to test this template.

File: examples/middleware_factory/src/combining_powertools_utilities_function.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+app_config = AppConfigStore(environment="dev", application="comments", name="features")
+feature_flags = FeatureFlags(store=app_config)
+
+
+@lambda_handler_decorator(trace_execution=True)
+def middleware_custom(handler: Callable, event: dict, context: LambdaContext):
+
+    # validating the INPUT with the given schema
+    # X-Customer-Id header must be informed in all requests
```
Comment (leandrodamascena) [inline examples/middleware_factory/src/combining_powertools_utilities_function.py:32]: Early return? I think I need other opinions here.

File: docs/utilities/middleware_factory.md [hunk 10]
```diff
@@ -3,89 +3,144 @@ title: Middleware factory
...
+### Tracing middleware **execution**
 
 If you are making use of [Tracer](../core/tracer.md), you can trace the execution of your middleware to ease operations.
 
 This makes use of an existing Tracer instance that you may have initialized anywhere in your code.
 
-```python hl_lines="3" title="Tracing custom middlewares with Tracer"
-from aws_lambda_powertools.middleware_factory import lambda_handler_decorator
+???+ warning
+    You must enable Active Tracing in your Lambda function when using this feature, otherwise Lambda will not be able to send traces to XRay..
```
Comment (rubenfonseca) [inline docs/utilities/middleware_factory.md]:
```suggestion
    You must [enable Active Tracing](../core/tracer/#permissions) in your Lambda function when using this feature, otherwise Lambda cannot send traces to XRay.
```

File: examples/middleware_factory/sam/combining_powertools_utilities_template.yaml [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+          Properties:
+            Path: /comments
+            Method: GET
+        GetCommentsById:
+          Type: Api
+          Properties:
+            Path: /comments/{comment_id}
+            Method: GET
+
+  # DYNAMODB TABLE FOR HISTORIC
```
Comment (rubenfonseca) [inline examples/middleware_factory/sam/combining_powertools_utilities_template.yaml]:
```suggestion
  # DynamoDB table to store historical data
```

File: examples/middleware_factory/sam/combining_powertools_utilities_template.yaml [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        - AttributeName: request_id
+          AttributeType: S
+      KeySchema:
+        - AttributeName: customer_id
+          KeyType: HASH
+        - AttributeName: request_id
+          KeyType: "RANGE"
+      BillingMode: PAY_PER_REQUEST
+
+  # APPCONFIG FOR FEATURE FLAGS
```
Comment (rubenfonseca) [inline examples/middleware_factory/sam/combining_powertools_utilities_template.yaml]:
```suggestion
  # Feature flags using AppConfig
```

File: examples/middleware_factory/src/combining_powertools_utilities_function.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+    # automatically adding security headers to all responses
+    # see: https://securityheaders.com/
+    logger.info("Injecting security headers")
+    response["headers"]["Referrer-Policy"] = "no-referrer"
+    response["headers"]["Strict-Transport-Security"] = "max-age=15552000; includeSubDomains; preload"
+    response["headers"]["X-DNS-Prefetch-Control"] = "off"
+    response["headers"]["X-Content-Type-Options"] = "nosniff"
+    response["headers"]["X-Permitted-Cross-Domain-Policies"] = "none"
+    response["headers"]["X-Download-Options"] = "noopen"
```
Comment (rubenfonseca) [inline examples/middleware_factory/src/combining_powertools_utilities_function.py:58]: oh boy this could totally be its own utility class for Powertools

File: examples/middleware_factory/src/getting_started_middleware_before_logic_function.py [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    ...
+
+
+@lambda_handler_decorator
+def middleware_before(handler, event, context) -> Callable:
+    # extract payload from a EventBridge event
+    detail: dict = extract_data_from_envelope(data=event, envelope=envelopes.EVENTBRIDGE)
+
+    # check if status_id exists in payload, otherwise add default state before processing payment
+    if not detail.get("status_id"):
```
Comment (rubenfonseca) [inline examples/middleware_factory/src/getting_started_middleware_before_logic_function.py]: still learning py, but wouldn't this be better?

```suggestion
    if "status_id" not in detail:
```

File: examples/middleware_factory/src/getting_started_middleware_with_params_function.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+
+@lambda_handler_decorator
+def obfuscate_sensitive_data(handler, event, context, fields: List) -> Callable:
+    # extracting payload from a EventBridge event
+    detail: dict = extract_data_from_envelope(data=event, envelope=envelopes.EVENTBRIDGE)
+    guest_data: Any = detail.get("guest")
+
+    # Obfuscate fields (email, vat, passport) before calling Lambda handler
+    if fields:
```
Comment (rubenfonseca) [inline examples/middleware_factory/src/getting_started_middleware_with_params_function.py]: maybe this condition is unecessary because the next line would just iterate over nothing

**Review by rubenfonseca — CHANGES_REQUESTED** [review]
Great job Leandro! Left some comments :)

File: examples/middleware_factory/src/getting_started_middleware_with_params_function.py (hunk 15 above)
Comment (leandrodamascena) [reply examples/middleware_factory/src/getting_started_middleware_with_params_function.py]: Agree 100%. I removed the condition

File: examples/middleware_factory/src/getting_started_middleware_before_logic_function.py (hunk 14 above)
Comment (leandrodamascena) [reply examples/middleware_factory/src/getting_started_middleware_before_logic_function.py]: You're right!

File: examples/middleware_factory/src/combining_powertools_utilities_function.py (hunk 13 above)
Comment (leandrodamascena) [reply examples/middleware_factory/src/combining_powertools_utilities_function.py:58]: Yea! We could add this as a utility and make things even safer!

**Review by rubenfonseca — APPROVED** [review]
LGTM
